### DISCUSSION TITLE
StrongSwan forwarding local networks of clients, VPS as a router?

### DISCUSSION THREAD

**Discussion by ettaluni** [strongswan/strongswan#254] [Q&A]
Good day. I am trying to communicate between two networks. One of the networks with a gray IP, so I bought a VPS server on Amazon, it also acts as a StronSwan server.
I did not have the opportunity to install StrongSwan on a router or gateway, so I set up two computers, on Debian 10, which will act as intermediaries and tunnel traffic. VPS server also on Debian 10.
Here is a brief description of the interfaces: 

	SERVER VPS:
	WAN Route: 18.18.18.18
	LAN: 172.26.13.11
	VPN IP: 10.1.22.0
	
	NODE A
	WAN Route: 91.91.91.91
	LAN: 192.168.0.82
	VPN IP: 10.1.22.1
	
	NODE B:
	WAN Route: 94.94.94.94
	LAN: 192.168.5.22
	VPN IP: 10.1.22.2

Graphical diagram of the tunnel that I need: in the attached files. 

I want to be able to access Local PC2 from Local PC1. That is, it turns out to be a rather difficult (for me) route.
Local PC1 (192.168.0.99) → Node A (192.168.0.82) (10.1.22.1) → VPS Server (172.26.13.11) (10.1.22.0) → Node B (192.168.5.22) (10.1.22.2) → Local PC2 ( 192.168.5.17).

At the moment, I was able to successfully set up a tunnel, from Node A (10.1.22.1) to Node B (10.1.22.2), but I don’t know how to forward the local network. 

I tried:
Added a lan-passthought connection because the client did not see its own local network and did not ping in any way.
Added a net-net connection, in the hope that it allowed traffic to be transmitted with the network address of another client. After that, the traffic began to go to the VPS server, which did not know what to do with it next.
Added a net-net connection to the Server-VPS to solve the problems described above, did not help.
Added ip xfrm rules to the VPS server and route, after which the VPS began to transmit traffic to another client, here they are: 

	ip route add 192.168.0.0/24 via 172.26.0.1 dev eth0 table 220
	ip xfrm policy add src 0.0.0.0/0 dst 192.168.0.0/24 dir out 	tmpl src 172.26.13.11 dst 91.91.91.91 proto esp reqid "1" mode 	tunnel
	ip xfrm policy add src 192.168.0.0/24 dst 0.0.0.0/0 dir in 	tmpl src 91.91.91.91 dst 172.26.13.11 proto esp reqid "1" mode 	tunnel
	ip xfrm policy add src 192.168.0.0/24 dst 0.0.0.0/0 dir fwd 	tmpl src 91.91.91.91 dst 172.26.13.11 proto esp reqid "1" mode 	tunnel

Actually, here I stopped, packets on the subnet of another client go through the tunnel to the Client, but he does not answer them. Doesn't even respond to own ip address (192.168.0.82)
The tcpdump screen is in the appendix. 

Actually that's the whole problem, I've been trying to raise the tunnel for several weeks now. 
![2021-03-12_16-48](https://user-images.githubusercontent.com/80467186/110949691-f2904f00-8353-11eb-8851-38df90f48087.png)
![2021-03-12_12-58](https://user-images.githubusercontent.com/80467186/110949693-f3c17c00-8353-11eb-82f6-c6ceed49b68d.png)
[NODE A-B.pdf](https://github.com/strongswan/strongswan/files/6130446/NODE.A-B.pdf)
[SERVER VPS.pdf](https://github.com/strongswan/strongswan/files/6130447/SERVER.VPS.pdf)

**Comment by Thermi** [comment]
Okay, easiest solution is to switch to an authentication method that uses raw ed25519 keys, drop the VIPs, and then just create two conns (you need them anyway), one for each PC that binds their corresponding identity to the right pair of leftsubnet and rightsubnet.

If you want to stick to EAP and VIPs, you need to do the following:

You need to create two different conns, one for each peer, with their corresponding local subnets as local_ts + dynamic. You need to apply the workaround from [Tobias' serverfault answer for eap-identity switching](https://serverfault.com/questions/908098/strongswan-clients-access-rights/908194#908194), because you need to configure two different pairs of remote_ts (rightsubnet) and local_ts (leftsubnet), because you need to allow access to other networks without NAT, + the assigned VIP.

In the end you need to end up with two conns where for each conn, rightid is the PC's identity (confirmed by the secret it uses) and the corresponding pair of leftsubnet and rightsubnet values to allow access to the corresponding subnets.

In any case, you need to make sure the traffic forwarded between the networks is not NATed by Netfilter (only happens if you configured your firewall to do so, even indirectly or unknowingly by using some firewall framework).

Are there any other requirements, like Internet access via the VPNs?

**Reply by ettaluni** [reply]
Thank you for answering, no Internet is needed, only sees specific computers through the tunnel. I changed the configuration according to Tobias' serverfault answer for eap-identity switching. Now I can ping Local PC1 and Local PC from the central VPS, but with Local PC1 I can't see Local PC2 and vice versa. According to my tcpdump checks. Packets go and redirect successfully: 

Local PC1(192.168.0.62) → Node A(192.168.0.82) → Server VPS(172.26.13.11) and back
Local PC2(192.168.5.17) → Node B(192.168.5.22) → Server VPS(172.26.13.11) and back

But the full route doesn't work:
Local PC1(192.168.0.62) → Node A(192.168.0.82) → Server VPS(172.26.13.11) хххх Node B(192.168.5.22)
Local PC2(192.168.5.17) → Node B(192.168.5.22) → Server VPS(172.26.13.11) хххх Node A(192.168.0.82)

And I already broke my head, tried many settings, leftsubnet - rightubnet. And it seems like strongsswan wrote all the routs himself. But the packages go nowhere. Here is the configuration, this time I did not write policy, route add table 220. I configured everything through the ipsec.conf configs.

[NODE A 16.03.2021.pdf](https://github.com/strongswan/strongswan/files/6154785/NODE.A.16.03.2021.pdf)
[NODE B 16.03.2021.pdf](https://github.com/strongswan/strongswan/files/6154786/NODE.B.16.03.2021.pdf)
[SERVER 16.03.2021.pdf](https://github.com/strongswan/strongswan/files/6154787/SERVER.16.03.2021.pdf)

**Reply by Thermi** [reply]
Hi, config looks good. How are you testing, and from which devices? Did you setup routes on both PCs to the remote network over the local nodes?

EDIT: I just saw you have compression enabled. If you want to pass that through Netfilter, you need to allow incoming ipip packets with a matching ipsec policy. That's because the compressed packets are first decapsulated (esp->ipip) and then decompressed (ipip->plain packet), going through Netfilter each time.
